core/src/main/java/com/sequenceiq/cloudbreak/util/StackUtil.java

```diff
@@ -158,13 +159,15 @@ public Set<Node> collectAndCheckReachableNodes(StackDto stack, Collection<String
     }
 
     public Set<Node> collectReachableNodes(StackDto stackDto) {
-        return hostOrchestrator.getResponsiveNodes(collectNodes(stackDto, emptySet()), gatewayConfigService.getPrimaryGatewayConfig(stackDto))
-                .getReachableNodes();
+        return hostOrchestrator.getResponsiveNodes(collectNodes(stackDto), gatewayConfigService.getPrimaryGatewayConfig(stackDto), false).getReachableNodes();
     }
 
-    public NodeReachabilityResult collectReachableAndUnreachableNodes(StackDto stack) {
-        NodeReachabilityResult nodeReachabilityResult = hostOrchestrator.getResponsiveNodes(collectNodes(stack, emptySet()),
-                gatewayConfigService.getPrimaryGatewayConfig(stack));
+    public NodeReachabilityResult collectReachableAndUnreachableNodes(StackDto stackDto, Optional<Set<String>> targets) {
+        Predicate<Node> nodePredicate = node -> targets.isEmpty() || targets.get().contains(node.getHostname());
+        NodeReachabilityResult nodeReachabilityResult = hostOrchestrator.getResponsiveNodes(
+                collectNodes(stackDto).stream().filter(nodePredicate).collect(Collectors.toSet()),
+                gatewayConfigService.getPrimaryGatewayConfig(stackDto),
+                targets.isPresent());
```

Comment:
If `targets` is not an Optional, `targeted` flag could be always true, since its called from targeted branch only.